syntax/encoding/object_test.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func some[T any](v T) *T {
+	return &v
+}
+
+func TestObjectBasic(t *testing.T) {
+	type Embedded struct {
+		Field string `syntax:"embedded"`
+	}
+
+	type basic struct {
+		Embedded
+
+		Null      any               `syntax:"null"`
+		Bool      bool              `syntax:"bool"`
+		Int       int               `syntax:"int"`
+		Int8      int8              `syntax:"int8"`
+		Int16     int16             `syntax:"int16"`
+		Int32     int32             `syntax:"int32"`
+		Int64     int64             `syntax:"int64"`
+		Uint      uint              `syntax:"uint"`
+		Uint8     uint8             `syntax:"uint8"`
+		Uint16    uint16            `syntax:"uint16"`
+		Uint32    uint32            `syntax:"uint32"`
+		Uint64    uint64            `syntax:"uint64"`
+		Float32   float32           `syntax:"float32"`
+		Float64   float64           `syntax:"float64"`
+		Number    json.Number       `syntax:"number"`
+		String    string            `syntax:"string"`
+		Array     []string          `syntax:"array"`
+		NilArray  []string          `syntax:"nilArray"`
+		Object    map[string]string `syntax:"object"`
+		NilObject map[string]string `syntax:"nilObject"`
+		Ptr       *string           `syntax:"ptr"`
+		Any       any               `syntax:"any"`
+	}
+
+	v := basic{
+		Embedded: Embedded{Field: "embedded-field"},
+		Null:     nil,
+		Bool:     true,
+		Int:      1,
+		Int8:     2,
+		Int16:    3,
+		Int32:    4,
+		Int64:    5,
+		Uint:     6,
+		Uint8:    7,
+		Uint16:   8,
+		Uint32:   9,
+		Uint64:   10,
+		Float32:  3.14,
+		Float64:  2.72,
+		Number:   json.Number("42"),
+		String:   "syntax",
+		Array:    []string{"hello", "world"},
+		Object:   map[string]string{"hello": "world"},
+		Ptr:      some("syntax"),
+		Any:      "value",
+	}
+
+	decoded, diags := DecodeValue(v)
+
+	expected := syntax.Object(
+		syntax.ObjectProperty(syntax.String("embedded"), syntax.String("embedded-field")),
+		syntax.ObjectProperty(syntax.String("bool"), syntax.Boolean(true)),
+		syntax.ObjectProperty(syntax.String("int"), syntax.Number(json.Number("1"))),
+		syntax.ObjectProperty(syntax.String("int8"), syntax.Number(json.Number("2"))),
+		syntax.ObjectProperty(syntax.String("int16"), syntax.Number(json.Number("3"))),
+		syntax.ObjectProperty(syntax.String("int32"), syntax.Number(json.Number("4"))),
+		syntax.ObjectProperty(syntax.String("int64"), syntax.Number(json.Number("5"))),
+		syntax.ObjectProperty(syntax.String("uint"), syntax.Number(json.Number("6"))),
+		syntax.ObjectProperty(syntax.String("uint8"), syntax.Number(json.Number("7"))),
+		syntax.ObjectProperty(syntax.String("uint16"), syntax.Number(json.Number("8"))),
+		syntax.ObjectProperty(syntax.String("uint32"), syntax.Number(json.Number("9"))),
+		syntax.ObjectProperty(syntax.String("uint64"), syntax.Number(json.Number("10"))),
+		syntax.ObjectProperty(syntax.String("float32"), syntax.Number(json.Number("3.140000104904175"))),
```

Comment:
The expected value could use some explanation. Is that a result of floating point error?

Reply:
Yeah 3.14 isn't representable as a float32, apparently.